src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs

```diff
@@ -193,6 +191,54 @@ bool IHttpBodyControlFeature.AllowSynchronousIO
             }
         }
 
+        PipeWriter IResponseBodyPipeFeature.ResponseBodyPipe
+        {
+            get
+            {
+                if (!object.ReferenceEquals(_cachedResponseBodyStream, ResponseBody))
```

Comment:
Just assign greedily like @davidfowl suggested. With this logic, if you set the body Stream first without then the body Pipe second without any calls to the getters in-between and then afterwards get the body Pipe, you'll get a Pipe wrapping the body Stream that was assigned first instead of the  body Pipe set after the body Stream was set.

We should have a test that assigns to both body properties and verifies the last assignment wins.

Reply:
Agreed. Done and added tests.